A few interface kernels, part 1 again

GiudGiud opened the pull request:
refs #17638

Those failing cases pass on my machine, so I m going to try a few things here

GiudGiud committed:
Use variable value in interfacepp, refs #17638
Recompute neighbor values for FV
Limit interface diffusive test to serialized mesh, due to interface mesh output limitations (see warning)
Limit number of processors on small 1D test to less than the number of elements
Remove lambda from the continuity test output. The values of the variables matters most

GiudGiud committed:
Revert "Merge pull request #17867 from aeslaughter/revert-17640"

This reverts commit d72a8d0d69e21b4945eedf2e78a7de80b1bd3e6f, reversing
changes made to 44a6a8b9b9a9f4c9185edab3efdbc3e0234363ae.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/742637/) on 3356dff wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/17870/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/17870/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format d72a8d0d69e21b4945eedf2e78a7de80b1bd3e6f`

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/742663/) on b890768 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/17870/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/742888/) on 7f9df5c wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/17870/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/17870/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format d72a8d0d69e21b4945eedf2e78a7de80b1bd3e6f`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/743563/) on f7da258 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/17870/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/17870/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format d72a8d0d69e21b4945eedf2e78a7de80b1bd3e6f`

GiudGiud commented:
@lindsayad should be good to go once test passes.
I need to learn how to read, the mesh output class is apparently not for distributed meshes.

GiudGiud commented:
Spoke too soon

GiudGiud commented:
Ok that was a trivial failure n_MPI > n_elements

Should be good now ^^

loganharbour commented:
> Ok that was a trivial failure n_MPI > n_elements

In almost all cases, this should't be a failure. Running 5 mpi procs on 4 elements should just leave a proc without elems which is very much valid.

GiudGiud commented:
Turns out it's not that.
It's another test that passes easily on my machine regardless of the number of procs, but smh on civet the lambda switches sign. 
I think the 'direction' of the interface might be getting flipped. I ll look deeper.

GiudGiud commented:
@lindsayad, should be good now?
I removed the lambda variable from the exodus, the values of u and v suffice for the test to show that it works properly

lindsayad commented:
Ok so this is the PR that I approved before right? I just looked at the new commit. Can you say what the previous problem was and how the new commit fixes them?

GiudGiud commented:
Yup
Those two lines say what I changed and why it passes the tests now:
-Limit interface diffusive test to serialized mesh, due to interface mesh output limitations (see warning)
-Remove lambda from the continuity test output. The values of the variables matters most

I'll expand on it:
- One of the interface diffusive tests was failing with 5+ processes. But it turns out the output class was returning a warning that the mesh should be serialized
- Another one was returning a diff at 15 and 16 processes. The lambda variable was inexplicably of an opposite sign. Since it does not matter what the lambda is as long as the variable values are good, I just removed it from the exodus. I dont really understand why it flipped sign. It feels like a elem_one_is_one kind of issue but did not find anything wrong with the code or the input. The residual that is likely flipped is var1 - var2, which does not care about elem_is_elem_one. 
-

lindsayad commented:
>     * But it turns out the postprocessor output class was returning a warning that the mesh should be serialized

Postprocessor output class? I'm not aware of any dedicated postprocessor output class. Are you talking about the Exodus output warning when running with distributed mesh? That warning should not be associated with Exodiffs.

>     * I dont really understand why it flipped sign. It feels like a elem_one_is_one kind of issue but did not find anything wrong with the code or the input

Element numbering will be different when doing replicated vs distributed meshes. E.g. which element is 1 will change

GiudGiud commented:
But var 1 and var 2 should not change.

lindsayad commented:
> Maybe it's exodiff. It's unrelated to my PR though

Can you paste the warning please

GiudGiud commented:
Warning at the top
```
postprocessors/interface_diffusive_flux.average/fv: Warning:  This MeshOutput subclass only supports meshes which have been serialized!
postprocessors/interface_diffusive_flux.average/fv: Warning:  This MeshOutput subclass only supports meshes which have been serialized!
postprocessors/interface_diffusive_flux.average/fv: 
postprocessors/interface_diffusive_flux.average/fv: Postprocessor Values:
postprocessors/interface_diffusive_flux.average/fv: +----------------+----------------+
postprocessors/interface_diffusive_flux.average/fv: | time           | diffusive_flux |
postprocessors/interface_diffusive_flux.average/fv: +----------------+----------------+
postprocessors/interface_diffusive_flux.average/fv: |   0.000000e+00 |   0.000000e+00 |
postprocessors/interface_diffusive_flux.average/fv: +----------------+----------------+
postprocessors/interface_diffusive_flux.average/fv: 
postprocessors/interface_diffusive_flux.average/fv:  0 Nonlinear |R| = 1.227562e+02
postprocessors/interface_diffusive_flux.average/fv:       0 Linear |R| = 1.227562e+02
postprocessors/interface_diffusive_flux.average/fv:       1 Linear |R| = 2.447205e+01
postprocessors/interface_diffusive_flux.average/fv:       2 Linear |R| = 1.334690e+01
postprocessors/interface_diffusive_flux.average/fv:       3 Linear |R| = 8.806458e+00
postprocessors/interface_diffusive_flux.average/fv:       4 Linear |R| = 6.210443e+00
postprocessors/interface_diffusive_flux.average/fv:       5 Linear |R| = 4.187028e+00
postprocessors/interface_diffusive_flux.average/fv:       6 Linear |R| = 2.456982e+00
postprocessors/interface_diffusive_flux.average/fv:       7 Linear |R| = 1.252794e+00
postprocessors/interface_diffusive_flux.average/fv:       8 Linear |R| = 4.806400e-01
postprocessors/interface_diffusive_flux.average/fv:       9 Linear |R| = 1.560912e-01
postprocessors/interface_diffusive_flux.average/fv:      10 Linear |R| = 4.610963e-02
postprocessors/interface_diffusive_flux.average/fv:      11 Linear |R| = 1.376783e-02
postprocessors/interface_diffusive_flux.average/fv:      12 Linear |R| = 5.607637e-03
postprocessors/interface_diffusive_flux.average/fv:      13 Linear |R| = 2.066341e-03
postprocessors/interface_diffusive_flux.average/fv:      14 Linear |R| = 7.691311e-04
postprocessors/interface_diffusive_flux.average/fv:  1 Nonlinear |R| = 7.691311e-04
postprocessors/interface_diffusive_flux.average/fv:       0 Linear |R| = 7.691311e-04
postprocessors/interface_diffusive_flux.average/fv:       1 Linear |R| = 3.594836e-04
postprocessors/interface_diffusive_flux.average/fv:       2 Linear |R| = 1.324549e-04
postprocessors/interface_diffusive_flux.average/fv:       3 Linear |R| = 6.437755e-05
postprocessors/interface_diffusive_flux.average/fv:       4 Linear |R| = 3.387056e-05
postprocessors/interface_diffusive_flux.average/fv:       5 Linear |R| = 1.918693e-05
postprocessors/interface_diffusive_flux.average/fv:       6 Linear |R| = 1.249788e-05
postprocessors/interface_diffusive_flux.average/fv:       7 Linear |R| = 8.568898e-06
postprocessors/interface_diffusive_flux.average/fv:       8 Linear |R| = 4.985107e-06
postprocessors/interface_diffusive_flux.average/fv:       9 Linear |R| = 2.707703e-06
postprocessors/interface_diffusive_flux.average/fv:      10 Linear |R| = 1.161607e-06
postprocessors/interface_diffusive_flux.average/fv:      11 Linear |R| = 4.116248e-07
postprocessors/interface_diffusive_flux.average/fv:      12 Linear |R| = 1.286030e-07
postprocessors/interface_diffusive_flux.average/fv:      13 Linear |R| = 3.647838e-08
postprocessors/interface_diffusive_flux.average/fv:      14 Linear |R| = 9.377859e-09
postprocessors/interface_diffusive_flux.average/fv:      15 Linear |R| = 2.111461e-09
postprocessors/interface_diffusive_flux.average/fv:  2 Nonlinear |R| = 2.111462e-09
postprocessors/interface_diffusive_flux.average/fv:  Solve Converged!
postprocessors/interface_diffusive_flux.average/fv: 
postprocessors/interface_diffusive_flux.average/fv: ===================================================================================
postprocessors/interface_diffusive_flux.average/fv: =   BAD TERMINATION OF ONE OF YOUR APPLICATION PROCESSES
postprocessors/interface_diffusive_flux.average/fv: =   PID 3028450 RUNNING AT meowbuild18.hpc.inl.gov
postprocessors/interface_diffusive_flux.average/fv: =   EXIT CODE: 139
postprocessors/interface_diffusive_flux.average/fv: =   CLEANING UP REMAINING PROCESSES
postprocessors/interface_diffusive_flux.average/fv: =   YOU CAN IGNORE THE BELOW CLEANUP MESSAGES
postprocessors/interface_diffusive_flux.average/fv: ===================================================================================
postprocessors/interface_diffusive_flux.average/fv: YOUR APPLICATION TERMINATED WITH THE EXIT STRING: Segmentation fault (signal 11)
postprocessors/interface_diffusive_flux.average/fv: This typically refers to a problem with your application.
postprocessors/interface_diffusive_flux.average/fv: Please see the FAQ page for debugging suggestions
postprocessors/interface_diffusive_flux.average/fv: 
postprocessors/interface_diffusive_flux.average/fv: 
postprocessors/interface_diffusive_flux.average/fv: Exit Code: 139
postprocessors/interface_diffusive_flux.average/fv: ################################################################################
postprocessors/interface_diffusive_flux.average/fv: Tester failed, reason: CRASH```

lindsayad commented on framework/src/fviks/FVScalarLagrangeMultiplierInterface.C:
```diff
@@ -0,0 +1,111 @@
+  mooseAssert((elem_dof_indices.size() == 1) && (neigh_dof_indices.size() == 1),
+              "We're currently built to use CONSTANT MONOMIALS");
+
+  // Primal
+  const auto primal_r = _lambda[0] * fi.faceArea() * fi.faceCoord();
+  _assembly.processDerivatives(primal_r, elem_dof_indices[0], _matrix_tags);
+  _assembly.processDerivatives(-primal_r, neigh_dof_indices[0], _matrix_tags);
```
This is the equation that will determine the sign of the LM variable. Which variable is the element variable will determine what the sign of `primal_r` needs to be and consequently what the lambda sign needs to be. You can imagine that if you swap which variable is element, the flux from 1 to 2 does not change, but the sign of `primal_r` and the LM must change. If you wanted to make this independent of processor count, then you'd have to multiply this `primal_r` by say 1 if `_elem_is_one` and -1 if not.

lindsayad replied:
I do agree that it is the primal variable values that matter, so I am fine with the solution of hiding the lambda output and not changing anything here.

GiudGiud replied:
This is the same as in FVInterfaceKernel though?
We look at which variable lives on elem 1, and add +residual to it.

I guess I had misunderstood it initially.

lindsayad replied:
In all our other flux objects a positive flux (e.g. a positive return quantity from `computeQpResidual()`) denotes a flux moving in the direction from `elem` to `neighbor` because that is how our normals are defined. Normals *must* be defined in a way that is irrespective of variables because it's a geometric feature. I think we want to continue to define a flux that is positive to be a flux that is moving in the normal direction.

GiudGiud replied:
Ok. Kind of having trouble thinking in terms of fluxes for the lagrange multiplier.
Either way, I will fix this, because it ll pop back up at the person making a penalty interface.

lindsayad commented:
> Warning at the top

Yea so what that means is that in order to generate output for Exodus we serialize the mesh on processor 0. This is a resource penalty, not an accuracy penalty. This should not change your results.

lindsayad commented:
So as far as I'm concerned:

>     * Another one was returning a diff at 15 and 16 processes. The lambda variable was inexplicably of an opposite sign. Since it does not matter what the lambda is as long as the variable values are good, I just removed it from the exodus. I dont really understand why it flipped sign. It feels like a elem_one_is_one kind of issue but did not find anything wrong with the code or the input. The residual that is likely flipped is var1 - var2, which does not care about elem_is_elem_one.

If this was settled by hiding the LM output, then I'm fine with that.

>     * One of the interface diffusive tests was failing with 5+ processes. But it turns out the postprocessor output class was returning a warning that the mesh should be serialized, so I just did that and that removed the issue since that test is not run anymore with those configurations.

But this issue doesn't seem to be resolved to my satisfaction

lindsayad commented:
Ok so the latter issue is the one where you get seg faults on CIVET but not on your own box? I could check this out on Rod if you want me to

lindsayad commented on framework/src/postprocessors/InterfaceDiffusiveFluxIntegral.C:
```diff
@@ -72,10 +66,9 @@ InterfaceDiffusiveFluxIntegralTempl<is_ad>::InterfaceDiffusiveFluxIntegralTempl(
     mooseWarning("Only one finite volume variable was specified. InterfaceDiffusiveFluxIntegral is "
-                 "only accurate at a "
-                 "FVDiffusionInterface, not with a regular diffusion kernel.");
+                 "only accurate at a FVDiffusionInterface, not with a regular diffusion kernel.");
```
I tend to think we should straight-up error instead of potentially producing inaccurate results

lindsayad commented on framework/src/postprocessors/InterfaceDiffusiveFluxIntegral.C:
```diff
@@ -0,0 +1,104 @@
+
+  // Warn that we are not using the same gradient to compute the diffusive flux here
+  if (_fv && !isParamValid("neighbor_variable"))
+    mooseWarning("Only one finite volume variable was specified. InterfaceDiffusiveFluxIntegral is "
+                 "only accurate at a FVDiffusionInterface, not with a regular diffusion kernel.");
```
I don't get why other conversation is outdated (I guess this filed moved?). But yea I think we should error here instead of potentially producing inaccurate results

GiudGiud replied:
I m not sure if it moved. 
But there is no other option right now for computing the diffusive flux at an interface

lindsayad replied:
I guess I don't understand combination of the conditional and the message itself. This only warns if you're doing FV and the user didn't specify a `neighbor_variable`. So my response to the *conditional* is "well if it's wrong to do FV without specifying a `neighbor_variable` then we should just error and force them to specify a `neighbor_variable`". Is my thinking wrong? Does "not with a regular diffusion kernel" mean not with a `FVDiffusion` kernel? Because that on its own makes me think you're talking about FE or something (only the fact that we only get here when `_fv` makes me think you're talking about `FVDiffusion`).

lindsayad replied:
Scratch that last question ... I can see that your concern is almost assuredly about having different forms for the diffusive flux, e.g. `FVDiffusion` includes non-orthgonal contributions whereas `FVDiffusionInterface` only does orthogonal contributions (e.g. difference in centroid values)

GiudGiud replied:
The user could be using the FVOneVarDiffusionInterface, or straight up FVDiffusion, and hoping to get the diffusive flux on an internal face. For the diffusive flux, we only have SideDiffusiveFlux and InterfaceDiffusiveFlux. The latter would be used in FE for the use cases mentioned, but in FV, we dont have an equivalent for them right now. 

It's not that inaccurate, it's still a centered approximation. I havent tested the order but I m going to claim second order right here right now ^^

GiudGiud replied:
Umm I may be misunderstanding lemme check

GiudGiud replied:
Ok the real problem is what kernels the user is using. This if statement is not catching all the bad options.

GiudGiud replied:
ah actually they are not. Well I ll try using it on rod and see what it does

lindsayad replied:
Yea I agree that the core concern is about what kernel object they are using. This object will not exactly reproduce the flux computed by a non-orthogonal kernel. So yea I think trying to determine whether you've got a good pairing is certainly outside the scope of a postprocessor constructor. I would say that as long as you have documented in the `.md` and perhaps even in the class description (which will show up in the `.md` generally) that this object will only *exactly* reproduce the fluxes for an orthgonal-based diffusion approximation (and I guess specifically `FVDiffusionInterface`), then I think you should drop the warning and go on your way

GiudGiud committed:
Address test failures:
 - make lagrange multiplier residual consistent when faceinfo element changes
 - skip computeQpIntegral in postprocessor if the faceInfo is not local
Change documentation links for FV interface kernels and interface postprocessor

Refs #17638

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/751794/) on c47a538 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/17870/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/17870/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 669aec3a443e9b20491caa79b4ae098d7c5ac9a0`

GiudGiud commented:
Ok they are diff-ing now with p5. I think there's something more at play. We are skipping values but no one is taking owneship, resulting in a lower diffusive flux than expected. 

There is nothing special about interface diffusive flux pp that is not in interface integral value, and apart from the 'interface' thing, the InternalVolumetricFlowRates and other internal face pps are also concerned by this issue of the ghosted face info. 

They all have an assert on fi not being null, so we're covered there in a way. But do we need the skip for everyone? It's what I m not sure.

lindsayad commented:
I think I'm going to have to dive into this a bit to understand why this isn't working now. I'm finishing up my CNSFV documentation but should be done in the next hour or so and then I'm going to pull this branch

lindsayad committed:
Be careful about when to excute interface pp when doing FV

lindsayad commented:
Putting this here for public record. This PR is illustrating that trying to put FV into our standard postprocessors is a disaster. Or put another way, having FaceInfo be its own thing really makes generic implementations that work for FE/FV extremely difficult. So here is a problem:
- the postprocessor element loop (ThreadedElementLoopBase) only goes into onInterface if there are sidesets associated with an elem-side pair. And a user might only specify a sideset on one side of an interface
- On the other hand, FaceInfo takes both sidesets associated with its elem-side pair and with its neighbor-neighor-side pair and plugs it into its boundaryIDs
- So: because of the first bullet if you just return if there is no FaceInfo and the user has specified only one sideset on the interface that happens to be on the neighbor side of the interface, then you will never count a diffusive flux on that face
- Now if you're clever in your postprocessor and say "hey, if I don't have a FaceInfo for the given elem-side pair, so I'm going to query for the neighbor pointer and figure out the neighbor side and then figure out if I have a FaceInfo for the neighbor-elem/neighbor-side pair", then you will double count interface fluxes if the user has happened to specify sidesets on both sides of the interface
- So I think we're going to have to do the fourth bullet *and* keep track of the face infos we've touched and then we should get things right

lindsayad commented:
Looks like you need to use `FVInterfaceKernels` instead of `fviks` in your `.md` pages

GiudGiud committed:
syntax calls in moosedocs needs the object class, not the folder, refs #17638

GiudGiud committed:
Move alex's fix to InterfaceUO, refs #17638

lindsayad commented on framework/src/userobject/InterfaceUserObject.C:
```diff
@@ -48,12 +48,58 @@ InterfaceUserObject::InterfaceUserObject(const InputParameters & parameters)
+
+  // Check for finite volume variables
+  _fv = false;
+  for (const auto & var : coupled_vars)
+    if (var->isFV())
+      _fv = true;
```
It seems like you are forcing a binary classification on something that might not be wholly finite volume or finite element. This seems somewhat trivial but maybe a rename from `_fv` to `_has_fv_vars` would help clarify this a bit?

GiudGiud committed:
Rename _fv to _has_fv_vars for interface UO, refs #17638

lindsayad reviewed (approved):
Nice work! Let's see if this can make it through round 2 of next -> devel testing 😄 Feel free to merge. The failure is unrelated to you

GiudGiud commented:
Thanks! Let's find out!
